mycpp/gc_str.cc

```diff
@@ -482,3 +485,97 @@ List<Str*>* Str::split(Str* sep) {
 
   return result;
 }
+
+// TODO: We should reduce the number of allocations and copies done here.
+static inline Str* _StrFromFormat(const char* fmt, va_list args) {
+  int n = strlen(fmt);
+  auto beg = std::cregex_iterator(fmt, fmt + n, gStrFmtRegex);
+  auto end = std::cregex_iterator();
+  mylib::BufWriter* w = Alloc<mylib::BufWriter>();
+  for (std::cregex_iterator match = beg; match != end; ++match) {
+    auto& lit = (*match)[1];
+    assert(lit.matched);
+    w->write(StrFromC(lit.str().c_str(), lit.str().size()));
+
+    bool zero_pad = false;
+    int width = 0;
+    auto& width_m = (*match)[2];
+    if (width_m.matched && !width_m.str().empty()) {
+      if (width_m.str()[0] == '0') {
+        zero_pad = true;
+        assert(width_m.str().size() > 1);
+        width = atoi(&width_m.str()[1]);
+      } else {
+        width = atoi(width_m.str().c_str());
+      }
+      assert(width >= 0 && width < kMaxFmtWidth);
+    }
+    std::ignore = zero_pad;  // XXX
+
+    char int_buf[kMaxFmtWidth];
+    Str* str_to_add = nullptr;
+    auto& code = (*match)[3];
+    if (!code.matched) {
+      assert(!width_m.matched);  // python errors on invalid format operators
+      goto done;
+    }
+    assert(code.str().size() == 1);
+    switch (code.str()[0]) {
+    case '%': {
+      assert(code.str().size() == 1);
+      str_to_add = StrFromC(code.str().c_str());
+      break;
+    }
+    case 's': {
+      str_to_add = va_arg(args, Str*);
+      zero_pad = false;  // python ignores the 0 directive for strings
+      break;
+    }
+    case 'r': {
+      str_to_add = va_arg(args, Str*);
+      str_to_add = repr(str_to_add);
+      zero_pad = false;  // python ignores the 0 directive for strings
+      break;
+    }
+    case 'd':  // fallthrough
+    case 'o': {
+      int d = va_arg(args, int);
+      assert(snprintf(int_buf, kMaxFmtWidth, match->str().c_str(), d) > 0);
```

Comment:
I was surprised to learn that `itoa()` et. al. aren't super portable, so we use `snprintf()` here. It feels a little awkward, though...